File: snowplow-tracker/src/androidTest/java/com/snowplowanalytics/snowplow/internal/tracker/MultipleInstancesTest.java
package com.snowplowanalytics.snowplow.internal.tracker;

import android.content.Context;

import androidx.test.ext.junit.runners.AndroidJUnit4;
import androidx.test.platform.app.InstrumentationRegistry;

import com.snowplowanalytics.snowplow.Snowplow;
import com.snowplowanalytics.snowplow.configuration.NetworkConfiguration;
import com.snowplowanalytics.snowplow.controller.TrackerController;

import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;

import java.util.Set;

import static org.junit.Assert.*;

@RunWith(AndroidJUnit4.class)
public class MultipleInstancesTest {

    @Before
    public void setUp() {
        Snowplow.removeAllTrackers();
    }

    @After
    public void tearDown() {
        Snowplow.removeAllTrackers();
    }

    @Test
    public void testSingleInstanceIsReconfigurable() {
        Context context = InstrumentationRegistry.getInstrumentation().getTargetContext();
        TrackerController t1 = Snowplow.createTracker(context,"t1", new NetworkConfiguration("snowplowanalytics.fake"));
        assertEquals("https://snowplowanalytics.fake/com.snowplowanalytics.snowplow/tp2", t1.getNetwork().getEndpoint());
        TrackerController t2 = Snowplow.createTracker(context,"t1", new NetworkConfiguration("snowplowanalytics.fake2"));
        assertEquals("https://snowplowanalytics.fake2/com.snowplowanalytics.snowplow/tp2", t2.getNetwork().getEndpoint());
        assertEquals(Set.of("t1"), Snowplow.getInstancedTrackerNamespaces());
        assertEquals(t1, t2);
    }

    @Test
    public void testMultipleInstances() {
        Context context = InstrumentationRegistry.getInstrumentation().getTargetContext();
        TrackerController t1 = Snowplow.createTracker(context,"t1", new NetworkConfiguration("snowplowanalytics.fake"));
        assertEquals("https://snowplowanalytics.fake/com.snowplowanalytics.snowplow/tp2", t1.getNetwork().getEndpoint());
        TrackerController t2 = Snowplow.createTracker(context,"t2", new NetworkConfiguration("snowplowanalytics.fake2"));
        assertEquals("https://snowplowanalytics.fake2/com.snowplowanalytics.snowplow/tp2", t2.getNetwork().getEndpoint());
        assertEquals(Set.of("t1", "t2"), Snowplow.getInstancedTrackerNamespaces());
        assertNotEquals(t1, t2);
    }

    @Test
    public void testDefaultTracker() {
        Context context = InstrumentationRegistry.getInstrumentation().getTargetContext();
        TrackerController t1 = Snowplow.createTracker(context,"t1", new NetworkConfiguration("snowplowanalytics.fake"));
        TrackerController t2 = Snowplow.createTracker(context,"t2", new NetworkConfiguration("snowplowanalytics.fake2"));
        TrackerController td = Snowplow.getDefaultTracker();
        assertEquals(t1, td);
    }

    @Test
    public void testUpdateDefaultTracker() {
        Context context = InstrumentationRegistry.getInstrumentation().getTargetContext();
        TrackerController t1 = Snowplow.createTracker(context,"t1", new NetworkConfiguration("snowplowanalytics.fake"));
        TrackerController t2 = Snowplow.createTracker(context,"t2", new NetworkConfiguration("snowplowanalytics.fake2"));
        Snowplow.setTrackerAsDefault(t2);
        TrackerController td = Snowplow.getDefaultTracker();
        assertEquals(t2, td);
    }

    @Test
    public void testRemoveTracker() {
        Context context = InstrumentationRegistry.getInstrumentation().getTargetContext();
        TrackerController t1 = Snowplow.createTracker(context,"t1", new NetworkConfiguration("snowplowanalytics.fake"));
        TrackerController t2 = Snowplow.createTracker(context,"t2", new NetworkConfiguration("snowplowanalytics.fake2"));
        Snowplow.removeTracker(t1);
        assertNotNull(t2);
        assertEquals(Set.of("t2"), Snowplow.getInstancedTrackerNamespaces());
    }

    @Test
    public void testRecreateTrackerWhichWasRemovedWithSameNamespace() {
        Context context = InstrumentationRegistry.getInstrumentation().getTargetContext();
        TrackerController t1 = Snowplow.createTracker(context,"t1", new NetworkConfiguration("snowplowanalytics.fake"));
        Snowplow.removeTracker(t1);
        TrackerController t2 = Snowplow.createTracker(context,"t1", new NetworkConfiguration("snowplowanalytics.fake2"));
        assertNotEquals(t1, t2);
        assertNotNull(t2);
        assertEquals(Set.of("t1"), Snowplow.getInstancedTrackerNamespaces());
    }

    @Test
    public void testRemoveDefaultTracker() {
        Context context = InstrumentationRegistry.getInstrumentation().getTargetContext();
        TrackerController t1 = Snowplow.createTracker(context,"t1", new NetworkConfiguration("snowplowanalytics.fake"));
        Snowplow.removeTracker(t1);
        TrackerController td = Snowplow.getDefaultTracker();
        assertNull(td);
        assertEquals(Set.of(), Snowplow.getInstancedTrackerNamespaces());
    }

    @Test
    public void testRemoveAllTrackers() {
        Context context = InstrumentationRegistry.getInstrumentation().getTargetContext();
        TrackerController t1 = Snowplow.createTracker(context,"t1", new NetworkConfiguration("snowplowanalytics.fake"));
        TrackerController t2 = Snowplow.createTracker(context,"t2", new NetworkConfiguration("snowplowanalytics.fake2"));
        Snowplow.removeAllTrackers();
        assertEquals(Set.of(), Snowplow.getInstancedTrackerNamespaces());
    }
}

File: snowplow-tracker/src/androidTest/java/com/snowplowanalytics/snowplow/internal/tracker/MultipleInstancesTest.kt
/*
 * Copyright (c) 2015-present Snowplow Analytics Ltd. All rights reserved.
 *
 * This program is licensed to you under the Apache License Version 2.0,
 * and you may not use this file except in compliance with the Apache License Version 2.0.
 * You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the Apache License Version 2.0 is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
 */
package com.snowplowanalytics.snowplow.internal.tracker

import androidx.test.ext.junit.runners.AndroidJUnit4
import androidx.test.platform.app.InstrumentationRegistry
import com.snowplowanalytics.snowplow.Snowplow.createTracker
import com.snowplowanalytics.snowplow.Snowplow.defaultTracker
import com.snowplowanalytics.snowplow.Snowplow.instancedTrackerNamespaces
import com.snowplowanalytics.snowplow.Snowplow.removeAllTrackers
import com.snowplowanalytics.snowplow.Snowplow.removeTracker
import com.snowplowanalytics.snowplow.Snowplow.setTrackerAsDefault
import com.snowplowanalytics.snowplow.configuration.NetworkConfiguration
import org.junit.After
import org.junit.Assert
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith

@RunWith(AndroidJUnit4::class)
class MultipleInstancesTest {
    @Before
    fun setUp() {
        removeAllTrackers()
    }

    @After
    fun tearDown() {
        removeAllTrackers()
    }

    @Test
    fun testSingleInstanceIsReconfigurable() {
        val context = InstrumentationRegistry.getInstrumentation().targetContext
        val t1 = createTracker(context, "t1", NetworkConfiguration("snowplowanalytics.fake"))
        Assert.assertEquals(
            "https://snowplowanalytics.fake/com.snowplowanalytics.snowplow/tp2",
            t1.network!!.endpoint
        )
        val t2 = createTracker(context, "t1", NetworkConfiguration("snowplowanalytics.fake2"))
        Assert.assertEquals(
            "https://snowplowanalytics.fake2/com.snowplowanalytics.snowplow/tp2",
            t2.network!!.endpoint
        )
        Assert.assertEquals(mutableSetOf("t1"), instancedTrackerNamespaces)
        Assert.assertEquals(t1, t2)
    }

    @Test
    fun testMultipleInstances() {
        val context = InstrumentationRegistry.getInstrumentation().targetContext
        val t1 = createTracker(context, "t1", NetworkConfiguration("snowplowanalytics.fake"))
        Assert.assertEquals(
            "https://snowplowanalytics.fake/com.snowplowanalytics.snowplow/tp2",
            t1.network!!.endpoint
        )
        val t2 = createTracker(context, "t2", NetworkConfiguration("snowplowanalytics.fake2"))
        Assert.assertEquals(
            "https://snowplowanalytics.fake2/com.snowplowanalytics.snowplow/tp2",
            t2.network!!.endpoint
        )
        Assert.assertEquals(mutableSetOf("t1", "t2"), instancedTrackerNamespaces)
        Assert.assertNotEquals(t1, t2)
    }

    @Test
    fun testDefaultTracker() {
        val context = InstrumentationRegistry.getInstrumentation().targetContext
        val t1 = createTracker(context, "t1", NetworkConfiguration("snowplowanalytics.fake"))
        createTracker(context, "t2", NetworkConfiguration("snowplowanalytics.fake2"))
        val td = defaultTracker
        Assert.assertEquals(t1, td)
    }

    @Test
    fun testUpdateDefaultTracker() {
        val context = InstrumentationRegistry.getInstrumentation().targetContext
        createTracker(context, "t1", NetworkConfiguration("snowplowanalytics.fake"))
        val t2 = createTracker(context, "t2", NetworkConfiguration("snowplowanalytics.fake2"))
        setTrackerAsDefault(t2)
        val td = defaultTracker
        Assert.assertEquals(t2, td)
    }

    @Test
    fun testRemoveTracker() {
        val context = InstrumentationRegistry.getInstrumentation().targetContext
        val t1 = createTracker(context, "t1", NetworkConfiguration("snowplowanalytics.fake"))
        val t2 = createTracker(context, "t2", NetworkConfiguration("snowplowanalytics.fake2"))
        removeTracker(t1)
        Assert.assertNotNull(t2)
        Assert.assertEquals(mutableSetOf("t2"), instancedTrackerNamespaces)
    }

    @Test
    fun testRecreateTrackerWhichWasRemovedWithSameNamespace() {
        val context = InstrumentationRegistry.getInstrumentation().targetContext
        val t1 = createTracker(context, "t1", NetworkConfiguration("snowplowanalytics.fake"))
        removeTracker(t1)
        val t2 = createTracker(context, "t1", NetworkConfiguration("snowplowanalytics.fake2"))
        Assert.assertNotEquals(t1, t2)
        Assert.assertNotNull(t2)
        Assert.assertEquals(mutableSetOf("t1"), instancedTrackerNamespaces)
    }

    @Test
    fun testRemoveDefaultTracker() {
        val context = InstrumentationRegistry.getInstrumentation().targetContext
        val t1 = createTracker(context, "t1", NetworkConfiguration("snowplowanalytics.fake"))
        removeTracker(t1)
        val td = defaultTracker
        Assert.assertNull(td)
        Assert.assertEquals(mutableSetOf<Any>(), instancedTrackerNamespaces)
    }

    @Test
    fun testRemoveAllTrackers() {
        val context = InstrumentationRegistry.getInstrumentation().targetContext
        createTracker(context, "t1", NetworkConfiguration("snowplowanalytics.fake"))
        createTracker(context, "t2", NetworkConfiguration("snowplowanalytics.fake2"))
        removeAllTrackers()
        Assert.assertEquals(mutableSetOf<Any>(), instancedTrackerNamespaces)
    }
}
